[OpenAI] Service version `v2023-12-01-preview` support (#37739)
Azure/azure-sdk-for-java: sdk/openai/azure-ai-openai/CHANGELOG.md

  mechanisms
- Early support for `gpt-4-vision-preview`, which allows the hybrid use of text and images as input to enable scenarios
  like "describe this image for me"
- Support for Azure enhancements to `gpt-4-vision-preview` results that include grounding and OCR features

### Breaking Changes

- Removed methods `getAudioTranscriptionAsResponseObject` and `getAudioTranscriptionAsResponseObjectWithResponse` from `OpenAIClient` and `OpenAIAsyncClient` classes.
Use `getAudioTranscription` or `getAudioTranscriptionWithResponse` convenience methods from respective classes.
- Removed methods `getAudioTranslationAsResponseObject` and `getAudioTranslationAsResponseObjectWithResponse` from `OpenAIClient` and `OpenAIAsyncClient` classes.
Use `getAudioTranslation` or `getAudioTranslationWithResponse` convenience methods from respective classes.

**`ChatMessage` changes:**

- The singular `ChatMessage` type has been replaced by `ChatRequestMessage` and `ChatResponseMessage`, the former of
  which is an abstract, polymorphic type with concrete derivations like `ChatRequestSystemMessage` and

>   `ChatRequestUserMessage`. This requires conversion from old `ChatMessages` into the new types. While this is

Reviewer: ```suggestion
  `ChatRequestUserMessage`. This requires conversion from old `ChatMessage` into the new types. While this is
```

  usually a straightforward string replacement, converting a response message into a request message (e.g. when
  propagating an assistant response to continue the conversation) will require creating a new instance of the
  appropriate request message with the response message's data. See the examples for details.

**Dall-e-3:**

- Azure OpenAI now uses `dall-e-3` model deployments for its image generation API and such a valid deployment must

>   be provided into the options for the `GetImageGenerations` method to receive results.

Reviewer: ```suggestion
  be provided into the options for the `getImageGenerations()` method to receive results.
```

Note: .NET method name starts with an uppercase letter but Java is lowercase.

### Features Added

> #### gpt-4-vision-preview

Reviewer: Maybe we can copy what .NET released CHANGELOG content. 

See https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/openai/Azure.AI.OpenAI/CHANGELOG.md#100-beta10-2023-12-06

Author: Sounds good! I took it instead from the PR description but this seems better!

#### Azure OpenAI On Your Data: new data sources and authentication mechanisms

Azure OpenAI On Your Data previously supported Azure Cognitive Search data sources via authentication with the administrator key for the ACS resource. This support is greatly expanded in this version:

| Data Source | Authentication Types |
|---|---|
| Azure Cognitive Search | API Key, System Managed Identity, User Managed Identity |
| Cosmos DB (MongoDB vCore) | Connection String |
| Pinecone | API Key |
| Elasticsearch | Key and Key ID |
| Azure Machine Learning Index | System Managed Identity, User Managed Identity, Access Token |
| Microsoft Search | Access Token |

> To use these new sources, provide an instance of the appropriate derived type of `OnYourDataChatExtensionConfiguration` for the desired data source and use a supported `OnYourDataAuthentication` instance for its `Authentication`.

Reviewer: ```suggestion
To use these new sources, provide an instance of the appropriate derived type of `OnYourDataVectorizationSource` for the desired data source and use a supported `OnYourDataAuthenticationOptions` instance for its `Authentication`.
```

#### dall-e-3 image generation

> Azure OpenAI now supports image generation with the improved `dall-e-3` models and this library supports the corresponding new features for quality level selection, prompt adjustment, and style. Use `OpenAIClient.GetImageGenerations()` and note that Azure OpenAI will *only* support `dall-e-3` models moving forward.

Reviewer: ```suggestion
Azure OpenAI now supports image generation with the improved `dall-e-3` models and this library supports the corresponding new features for quality level selection, prompt adjustment, and style. Use `getImageGenerations()` and note that Azure OpenAI will *only* support `dall-e-3` models moving forward.
```

Use `getAudioTranslation` or `getAudioTranslationWithResponse` convenience methods from respective classes.

> - Removed property, ResponseError `error` from `ImageLocation` class.

Reviewer: ```suggestion
```
